magento/loaders/productDetailsPage.ts

```diff
@@ -0,0 +1,168 @@
+}
+
+/**
+ * @title VNDA Integration
+ * @description Product Details Page loader
+ */
+async function loader(
+  props: Props,
+  req: Request,
+  ctx: AppContext,
+): Promise<ProductDetailsPage | null> {
+  console.log("loader start");
+  const url = new URL(req.url);
+  const { slug, currencyCode = "" } = props;
+  const { clientGuest, clientAdmin, site, storeId } = ctx;
+
+  if (!slug) {
+    return null;
+  }
+
+  const getMaybeProduct = async (slug: string) => {
+    try {
+      const response = await clientAdmin["GET /rest/:site/V1/products"]({
+        site,
+        currencyCode: currencyCode,
+        [KEY_FIELD]: URL_KEY,
+        [KEY_VALUE]: slug,
+      }).then((res) => res.json());
+      return response.items[0];
+    } catch (_error) {
+      return null;
+    }
+  };
+  const getProductPriceAndImages = async (sku: string | null) => {
+    if (!sku) {
+      return null;
+    }
+    try {
+      const response = await clientGuest
+        ["GET /rest/:site/V1/products-render-info"]({
+          site,
+          [KEY_FIELD]: "sku",
+          [KEY_VALUE]: sku,
+          storeId: storeId,
+          currencyCode: currencyCode,
+          fields: "items[price_info,image,currency_code,url]",
+        }).then((res) => res.json());
+
+      return {
+        priceInfo: response.items[0].price_info,
+        currencyCode: response.items[0].currency_code,
+        images: response.items[0].images,
+        url: response.items[0].url,
+      };
+    } catch (_error) {
+      return null;
+    }
+  };
+  const getProductStock = async (sku: string | null) => {
+    if (!sku) {
+      return null;
+    }
+    try {
+      const response = await clientGuest
+        ["GET /rest/:site/V1/products-render-info"]({
+          site,
+          [KEY_FIELD]: "sku",
+          [KEY_VALUE]: sku,
+          storeId: storeId,
+          currencyCode: currencyCode,
+          fields: "items[extension_attributes.stock_item]",
+        }).then((res) => res.json());
+      return response.items[0].extension_attributes?.stock_item;
+    } catch (_error) {
+      return null;
+    }
+  };
+  const getCategoryName = async (categoryId: string) => {
+    try {
+      return await clientAdmin["GET /rest/:site/V1/categories/:categoryId"]({
+        site,
+        categoryId,
+        fields: "name,position",
+      }).then((res) => res.json());
+    } catch (error) {
+      return error;
+    }
+  };
+
+  const productInfo = await getMaybeProduct(slug);
+
+  // 404: product not found
+  if (!productInfo) {
+    console.log("product not found");
+    return null;
+  }
+
+  const [productPriceInfos, stock_item] = await Promise.all(
+    [
+      getProductPriceAndImages(productInfo.sku),
+      getProductStock(productInfo.sku),
+    ],
+  );
+
+  // 404: price not found
+  if (!productPriceInfos) {
+    return null;
+  }
+
+  const categoryLinks = productInfo.extension_attributes?.category_links;
+
+  const product = toProduct({
+    ...productInfo!,
```

Comment:
Não deveria precisar de ! aqui
```suggestion
    ...productInfo,
```